A71-xds-fallback.md

```diff
@@ -21,81 +21,79 @@ obtained.
-evolving into an industry standard. gRPC introduced a number of features that
-can be configured using this protocol.
-
-Components of the server mesh communicate with xDS server(s) (also known as
-a "control plane") to obtain the initial configuration and to receive updates
-to the configuration.
-
-An xDS configuration consists of a number of "resources" that can be queried
-separately and may change independently from one another.
-
-Even a brief downtime of the xDS control plane may cause significant disruption
-in the service mesh inter-component connectivity and result in wider outages.
-Examples include:
-
-* Services becoming unavailable as they fail to access downstream dependencies.
-* Traffic being routed incorrectly failing to properly balance the load between
-the endpoints.
-* Security risks may be introduced as outdated credentials are used for
-communications between the servers.
+and subscribing to the configuration of a server mesh. Even a brief downtime
+of the xDS control plane may cause significant disruption in the service mesh
+inter-component connectivity and result in wider outages.
 
 Current xDS implementation in gRPC uses a single global instance of
 the XdsClient class to communicate with the xDS control plane. This instance
 is shared across all channels to reduce overhead by providing a shared cache
-for xDS resources as well as reducing a number of connections to xDS servers.
+for xDS resources as well as reducing the number of connections to xDS servers.
 
 ### Related Proposals: 
-* [A27: xDS-Based Global Load Balancing](A27)
-* [A36: xDS-Enabled Servers](A36)
-* [A57: XdsClient Failure Mode Behavior](A57)
+* [A27: xDS-Based Global Load Balancing][A27]
+* [A36: xDS-Enabled Servers][A36]
+* [A47: xDS Federation][A47]
+* [A57: XdsClient Failure Mode Behavior][A57]
 
 [A27]: A27-xds-global-load-balancing.md
 [A36]: A36-xds-for-servers.md
+[A47]: A47-xds-federation.md
 [A57]: A57-xds-client-failure-mode-behavior.md
 
 ## Proposal
 
-gRPC code should use fallback iff both:
+gRPC code should use fallback if the following conditions hold:
 
-* primary server not reachable
-* at least one watcher exists for a resource that is not cached (where
-   "does not exist" counts as cached).
+* There had been a failure during resource retrieval, as described in [A57]:
+    - Connection to the data plane fails. 
+    - The ADS stream being closed before the first server response had been
+      received.
+* At least one watcher exists for a resource that is not cached. Resources are
+    considered cached if:
+    - A resource had been successfully retried from the xDS server.
+    - Server reported the resource had been found.
+    - An error was reported retrieving the resource over a connection that had
+      been successfully established beforehand (as described above). In C-core
+      such resource will have a `REQUESTED` state.
 
 Instead of using a global XdsClient instance, gRPC will use a shared XdsClient
```

Comment:
Note that switching from a singleton shared XdsClient to a separate shared XdsClient per channel target has an implication for CSDS (see [gRFC A40](https://github.com/grpc/proposal/blob/master/A40-csds-support.md), which should also be listed in the "related proposals" section above).  Specifically, instead of CSDS reporting a single global state for each resource, it will instead need to report the state on a per-channel-target basis.  In most cases, this won't matter, since the set of xDS resources used for a given channel target will often be disjoint, but there could be cases where a given resource is shared between channels.  And in that case, each of those channels may have a different state for the resource, so they need to be reported separately.

To support this, we'll need to propose an API change for CSDS to add a field to the [`ClientConfig` message in the CSDS response](https://github.com/envoyproxy/envoy/blob/d99027a8bf61f1ddfc498310aabb79092ef1b1b2/api/envoy/service/status/v3/csds.proto#L130) to indicate which channel target the data is associated with.